shell/platform/android/external_view_embedder/surface_pool.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_ANDROID_EXTERNAL_VIEW_EMBEDDER_SURFACE_POOL_H_
+#define FLUTTER_SHELL_PLATFORM_ANDROID_EXTERNAL_VIEW_EMBEDDER_SURFACE_POOL_H_
+
+#include "flutter/flow/surface.h"
+#include "flutter/shell/platform/android/context/android_context.h"
+#include "flutter/shell/platform/android/surface/android_surface.h"
+
+namespace flutter {
+
+//------------------------------------------------------------------------------
+/// An Overlay layer represents an `android.view.View` in the C side.
+///
+/// The `id` is used to uniquely identify the layer and recycle it between
+/// frames.
+///
+struct OverlayLayer {
+  OverlayLayer(int id,
+               std::unique_ptr<AndroidSurface> android_surface,
+               std::unique_ptr<Surface> surface);
+
+  ~OverlayLayer();
+
+  // A unique id to identify the overlay when it gets recycled.
+  const int id;
+
+  // A GPU surface.
+  const std::unique_ptr<AndroidSurface> android_surface;
+
+  // A GPU surface. This may change when the overlay is recycled.
+  std::unique_ptr<Surface> surface;
+
+  // The `GrContext` that is currently used by the overlay surfaces.
+  // We track this to know when the GrContext for the Flutter app has changed
+  // so we can update the overlay with the new context.
+  //
+  // This may change when the overlay is recycled.
+  GrContext* gr_context;
+};
+
+// This class isn't thread safe.
+class SurfacePool {
```

Comment:
As described in the previous comment, this interface can be reworked to remove layers that have not been reused.